### DISCUSSION TITLE
How to create animated GIF  from multiple VImage loaded from pngs?

### DISCUSSION THREAD

**Discussion by lanyusan** [libvips/libvips#3146] [Help wanted]
I read the documentation about functions for GIFs. It isn't very clear to me if libvips can concatenate multiple single frame images like pngs into a multiple frame gif, with each frame its own delays.

I am currently using imagesharp. 

Thanks in advance. 

I have been using sharp on nodejs for a long time. Never had a chance to say thanks to libvips maintainers directly. It is a great piece of work.

**Comment by jcupitt** [comment]
Hi @lanyusan,

You combine the frames vertically into something like a film strip, then save as GIF, setting the page height to the vertical frame size. You attach an array of delay times as a metadata item.

To split an animation into frames in python, perhaps:

```python
#!/usr/bin/python3

import sys
import pyvips

# load the animation, chop into pages
image = pyvips.Image.new_from_file(sys.argv[2], n=-1, access="sequential")
page_height = image.get("page-height")
n_pages = image.get("n-pages")
delay = image.get("delay")
print(f"delay array = {delay}")

pages = [image.crop(0, page_number * page_height, image.width, page_height)
         for page_number in range(0, n_pages)]
print(f"writing frames to {sys.argv[1]} ...")
for page_number in range(len(pages)):
    filename = f"{sys.argv[1]}/frame-{page_number:04}.png"
    pages[page_number].write_to_file(filename)
```

And to join again, maybe:

```python
#!/usr/bin/python3

import sys
import pyvips

# the input images
# assume these are all the same size
images = [pyvips.Image.new_from_file(filename, access="sequential")
          for filename in sys.argv[2:]]

# frame delays are in milliseconds
delay_array = [300] * len(images)

animation = pyvips.Image.arrayjoin(images, across=1)
animation.set_type(pyvips.GValue.gint_type, "page-height", images[0].height)
animation.set_type(pyvips.GValue.array_int_type, "delay", delay_array)
print(f"writing {sys.argv[1]} ...")
animation.write_to_file(sys.argv[1])
```

I see:

```
$ mkdir frames
$ ./split-animation.py frames ~/pics/3198.gif
delay array = [30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30]
writing frames to frames ...
$ ./join-animation.py x.gif frames/*
writing x.gif ...
$
```

On this PC, it's:

```
$ vipsheader frames/frame-0000.png 
frames/frame-0000.png: 298x193 uchar, 3 bands, srgb, pngload
$ /usr/bin/time -f %M:%e ./join-animation.py x.gif frames/*
writing x.gif ...
69592:1.24
```

So 70mb of memory and 1.25s to make a 138 frame 300x200 pixel animation. I don't know how that compares to imagesharp, you'd need to do some benchmarking.

Some of that time is spent in PNG decompress, it comes down to about 1s if you use TIFF to store the frames.

**Comment by jcupitt** [comment]
> I have been using sharp on nodejs for a long time. Never had a chance to say thanks to libvips maintainers directly. It is a great piece of work.

Oh no problem. I'm glad it's useful!

**Reply by lanyusan** [reply]
By the way, when I tried to run net-vips version in multiple treads, I got this error:

```
(process:120518): GLib-GObject-CRITICAL **: 23:34:42.856: g_object_ref: assertion 'G_IS_OBJECT (object)' failed
```

In single thread, all is good.

Does libvips support mutiple threads accessing the same vimage buffer?

Thanks.

**Reply by jcupitt** [reply]
Yes, it should be fine, I think. Could you make a minimal example and post to the net-vips issue tracker?

**Comment by lanyusan** [comment]
Thanks @jcupitt 

I have got it work with net-vips and benchmarked against imagesharp.

Libvips has much smaller output size and also better colors while imagesharp has faster speed. Both were ran with default config.

It took libvips 4.5 minutes while imagesharp 0.8 minutes. Obviously it is comparing apple to orange in terms of speed as level of optimization between the two were vastly different.

I would certainly tune both to meet in the middle in terms of optimization to see how it goes.

libvisp samples :

390k & 287k

![100](https://user-images.githubusercontent.com/56706512/200348488-2eaec8ee-2ea6-4067-8ac5-cbb17452d946.gif)
![1](https://user-images.githubusercontent.com/56706512/200348638-1b5ea45c-93b9-4f34-a34c-71755db07dd7.gif)


imagesharp samples:

1.1m & 1.1m

![100](https://user-images.githubusercontent.com/56706512/200348299-eeea6fbe-5fd4-4a25-a679-a96ad3f7fefa.gif)
![1](https://user-images.githubusercontent.com/56706512/200348381-61a077fb-4b6e-4194-a918-695bb1823fe0.gif)

**Reply by jcupitt** [reply]
Yes, different development priorities I guess. Thanks for posting the benchmarks!

**Reply by JimBobSquarePants** [reply]
Does libvips dither the output? You can see the telltale pixel shaking caused by dithering in the ImageSharp output.

**Reply by lanyusan** [reply]
> Does libvips dither the output? You can see the telltale pixel shaking caused by dithering in the ImageSharp output.

Yes i think so.

https://www.[libvips](https://www.libvips.org/API/current/VipsForeignSave.html#vips-gifsave).org/API/current/VipsForeignSave.html#vips-gifsave

**Reply by jcupitt** [reply]
Exactly, by default libvips uses the high quality dither library from pngquant:

https://github.com/ImageOptim/libimagequant

Their dither is good -- noticeably better than the one in imagemagick, for example.

**Reply by JimBobSquarePants** [reply]
Ah yes, it's a modified Floyd Steinberg with clipping to avoid phantom pixels. Impossible to beat IMO while maintaining flexibility to use other dither options (_ImageSharp supports bring your own plus you can scale the strength between 0-1_). The quantizer itself is great also.

**Reply by jcupitt** [reply]
Yes, it's a nice thing. libvips exposes a 0-1 "dither strength" control too, and also uses the pngquant palette optimiser.

darthsim made a similar thing:

https://github.com/DarthSim/quantizr

Which libvips will use if libimagequant is not around.

**Reply by lanyusan** [reply]
Hi @jcupitt, 
I am trying to optimize gif encoding by using disposal method. 

How to set it with libvips? 

Is it similar to set delays ? I have tried out the below,which seems not working.

Ps. checked cgif wiki. It seems it doesn't support disposal method.

https://github.com/dloebl/cgif/wiki/General-API#cgif_addframe

```
                using var frames = framesJoined.Mutate(mutable =>
                {
                    // Set the number of loops, libvips uses iterations like this:
                    // 0 - set 0 loops (infinite)
                    // 1 - loop once
                    // 2 - loop twice etc.
                    mutable.Set(GValue.GIntType, "loop", 0);

                    // Set the frame delay(s)
                    // We use a 1000ms delay for all frames in this example
                    mutable.Set(GValue.ArrayIntType, "delay", delayArr);

                    // Set page height
                    mutable.Set(GValue.GIntType, "page-height", height);
                   // is this the way to set disposal?
                    mutable.Set(GValue.ArrayIntType, "disposal", disposalArr); // first frame dispose
                });
```

**Reply by kleisauke** [reply]
AFAIK, controlling the disposal method can't be done currently, since cgif uses that to optimize GIF images. 
https://github.com/libvips/libvips/blob/557fc76634a60a0fb1c4a0fca890af3570d98829/libvips/foreign/cgifsave.c#L531-L536

**Comment by dloebl** [comment]
Hello @lanyusan,
> I am trying to optimize gif encoding by using disposal method.

As Kleis already mentioned, [cgif](https://github.com/dloebl/cgif) uses the disposal method to optimize the file size of animated GIFs.
It isn't possible to directly specify the disposal method in [cgif](https://github.com/dloebl/cgif) at the moment. Do you want to optimize the file size or the encoding speed?

**If it's about the file size:**
A good benchmark for the file size of GIFs is [gifsicle](https://github.com/kohler/gifsicle), as it supports some optimizations [cgif](https://github.com/dloebl/cgif) doesn't support yet. With the GIF you provided ([test.gif](https://user-images.githubusercontent.com/56706512/200348488-2eaec8ee-2ea6-4067-8ac5-cbb17452d946.gif)) I see:
```
$ gifsicle -O3 test.gif -o opt.gif
$ du -h --si test.gif opt.gif
394k	test.gif
377k	opt.gif
```


**A couple of other recommendations to improve the file size of GIFs:**
1. Smaller bitdepths (usually) require less bits per pixel - however, dithering is going to be more visible:
```
$ vips copy test.gif[n=-1] bitdepth6.gif[bitdepth=6]
$ du -h --si bitdepth6.gif
304k	bitdepth6.gif
```
2. [libvips v8.13](https://www.libvips.org/2022/05/28/What's-new-in-8.13.html) improved `gifsave` in general and added some parameters which can decrease the file size further: https://www.libvips.org/API/current/VipsForeignSave.html#vips-gifsave
`interframe_maxerror` and `interpalette_maxerror`:
```
$ vips copy test.gif[n=-1] maxerror8.gif[interframe_maxerror=8.0]
$ du -h --si maxerror8.gif
361k	maxerror8.gif
```
These parameters are lossy, so you would want to be careful with them.

| original | bitdepth 6 | maxerror 8 |
| - | - | - |
| ![test](https://user-images.githubusercontent.com/34104349/200668103-4b0e1d89-6032-430e-8fb8-de8d7044450f.gif) | ![depth6](https://user-images.githubusercontent.com/34104349/200668146-49f4a14b-e894-4e8e-a8bb-4d54e75a93a7.gif) | ![maxerror8](https://user-images.githubusercontent.com/34104349/200668176-0b6c71b5-c8f4-42bc-bf63-a04437845ad7.gif) |
| 394 KB | 304 KB | 361 KB |

**Reply by lanyusan** [reply]
@dloebl Thanks for the reply.

Right now I want to optimize encoding speed. The file size and color sampling quality of libvips are already very satisfying.

The project I am working on sometimes needs to generate large amount of GIFs, up to 100,000 with each around 50 frames and dimensions of up to 1000 x 1000.

**Reply by lanyusan** [reply]
@dloebl  @jcupitt  I have tried some combinations of dither and effort for this function:
```
int
vips_gifsave (*in,
              const char *filename,
              ...);
```

And I found out `0.5 1` is a pretty good combination of quality and speed.

I could generate and optimize 100 gifs with 4 threads in 2minutes. Output quality is pretty good and file size is just around 10% larger than more aggressive optimization.

I am very happy with it.

Thank you all for all the help.
